beacon/lib/beacon/supervisor.ex

```diff
@@ -0,0 +1,53 @@
+defmodule Beacon.Supervisor do
+  @moduledoc false
+  use Supervisor
+
+  def name(scope), do: :"#{scope}_beacon"
+  def supervisor_name(scope), do: :"#{scope}_beacon_supervisor"
+  def partition_name(scope, partition), do: :"#{scope}_beacon_partition_#{partition}"
+
+  @spec partition(atom, Scope.group()) :: atom
+  def partition(scope, group) do
+    case :persistent_term.get(scope, :unknown) do
+      :unknown -> raise "Beacon for scope #{inspect(scope)} is not started"
+      partition_names -> elem(partition_names, :erlang.phash2(group, tuple_size(partition_names)))
+    end
+  end
+
+  @spec partitions(atom) :: [atom]
+  def partitions(scope) do
+    case :persistent_term.get(scope, :unknown) do
+      :unknown -> raise "Beacon for scope #{inspect(scope)} is not started"
+      partition_names -> Tuple.to_list(partition_names)
+    end
+  end
+
+  @spec start_link(atom, pos_integer(), Keyword.t()) :: Supervisor.on_start()
+  def start_link(scope, partitions, opts \\ []) do
+    args = [scope, partitions, opts]
+    Supervisor.start_link(__MODULE__, args, name: supervisor_name(scope))
+  end
+
+  @impl true
+  def init([scope, partitions, opts]) do
+    children =
+      for i <- 0..(partitions - 1) do
+        partition_name = partition_name(scope, i)
+
+        ^partition_name =
+          :ets.new(partition_name, [:set, :public, :named_table, read_concurrency: true])
+
+        %{id: i, start: {Beacon.Partition, :start_link, [scope, partition_name]}}
```

Comment:
I understand the idea here is to not lose data if the partition process crashes but I am not sure if I would do this. First of all, the process is not meant to crash. Second, if the process crashes, you don't know if the root cause was because the ETS table got corrupted! So reusing the ETS table will now make this process constantly crash, and eventually the whole node to blow up. IIRC, the Registry links itself to monitored processes, so in case there is a bug in the Registry, every process it is "watching" crashes too.